tools/isobuild/compiler-plugin.js

```diff
@@ -827,20 +832,141 @@ class ResourceSlot {
 
   addError(message, info) {
     // If this file is ever actually imported, only then will we report
-    // the error. Use this.jsOutputResources because that's what the
-    // ImportScanner deals with.
-    this.jsOutputResources.push({
-      type: "js",
-      sourcePath: this.inputResource.path,
-      targetPath: this.inputResource.path,
-      servePath: this.inputResource.path,
-      data: Buffer.from(
-        "throw new Error(" + JSON.stringify(message) + ");\n",
-        "utf8"),
-      lazy: true,
-      error: { message, info },
+    // the error.
+    this.errors.push({ message, info });
+  }
+}
+
+class OutputResource {
+  constructor({
+    type,
+    resourceSlot,
+    options = Object.create(null),
+    lazyFinalizer = null,
+  }) {
+    this._lazyFinalizer = lazyFinalizer;
+    this._initialOptions = options;
+    this._finalizerPromise = null;
+    // Share the errors array of the resourceSlot.
+    this._errors = resourceSlot.errors;
+
+    let sourcePath = resourceSlot.inputResource.path;
+    if (_.has(options, "sourcePath") &&
+        typeof options.sourcePath === "string") {
+      sourcePath = options.sourcePath;
+    }
+
+    const targetPath = options.path || sourcePath;
+
+    Object.assign(this, {
+      type,
+      lazy: resourceSlot._isLazy(options, true),
+      bare: !! resourceSlot._getOption("bare", options),
+      mainModule: !! resourceSlot._getOption("mainModule", options),
+      sourcePath,
+      targetPath,
+      servePath: resourceSlot.packageSourceBatch
+        .unibuild.pkg._getServePath(targetPath),
     });
   }
+
+  finalize() {
+    if (this._finalizerPromise) {
+      this._finalizerPromise.await();
+    } else if (this._lazyFinalizer) {
+      const finalize = this._lazyFinalizer;
+      this._lazyFinalizer = null;
+      (this._finalizerPromise = new Promise(
+        resolve => resolve(finalize())
+      ).then(result => {
+        Object.assign(this._initialOptions, result);
+        this._finalizerPromise = null;
+      })).await();
```

Comment:
It's too bad this implementation has to use `Fiber`s to await the result of the `lazyFinalizer` function, but that's an implementation detail we can revisit in future versions of Meteor without changing the API.